### PR TITLE
Extend storage status

### PR DESCRIPTION
**Pull request by chfast**
Expand the number of storage statuses to cover all possible EVM
behaviors. These provide enough information to VM to compute not only
gas costs but also refunds. Including legacy SSTORE costs before net
gas metering.

### CODE REVIEW

**Comment by chfast** [comment]
I have Client-side implementation of this using only 4 checks:
- `original != current` (dirty)
- `original == new_value` (restored)
- `current != 0`
- `new_value != 0`

https://github.com/ethereum/evmone/blob/state/test/state/host.cpp#L99-L162 (unstable link).

Then you can get the gas cost and refund per revision from a lookup table.

File: include/evmc/evmc.h [hunk 1]
```diff
@@ -503,36 +503,111 @@ typedef evmc_bytes32 (*evmc_get_storage_fn)(struct evmc_host_context* context,
...
+     * in the transaction context. This is the group of "no-op" cases related to
+     * minimal gas cost.
+     * 0|X   -> 0 -> 0 (current value unchanged)
+     * 0|X|Y -> Y -> Y (current value unchanged)
+     * 0|X   -> Y -> Z (modified previously added/modified value)
+     *
+     * This corresponds to the EIP-2200 rule 2 where:
+     * c == v.
+     */
+    EVMC_STORAGE_MODIFIED_AGAIN = 0,
```
Comment (yperbasis) [inline include/evmc/evmc.h:530]: Why `EVMC_STORAGE_MODIFIED_AGAIN` rather than `EVMC_STORAGE_UNCHANGED`?

Comment (chfast) [reply include/evmc/evmc.h:530]: It felt more correct to me: in general we modify a value, just gas cost does not care. The "unchanged" strongly suggests that `current == value` what is not always the case for this group. But I can change it to `EVMC_STORAGE_UNCHANGED` if people think this is fine.

File: include/evmc/evmc.h [hunk 2]
```diff
@@ -503,36 +503,111 @@ typedef evmc_bytes32 (*evmc_get_storage_fn)(struct evmc_host_context* context,
...
+ *
+ * The order of elements follows EIPs introducing net storage gas costs:
+ * - EIP-2200: https://eips.ethereum.org/EIPS/eip-2200,
+ * - EIP-1283: https://eips.ethereum.org/EIPS/eip-1283.
  */
 enum evmc_storage_status
 {
     /**
-     * The value of a storage item has been left unchanged: 0 -> 0 and X -> X.
+     * The value of a storage item has been left unchanged has been modified again
```
Comment (chfast) [inline include/evmc/evmc.h:520]: "or" missing.

File: include/evmc/evmc.h (hunk 1 above)
Comment (yperbasis) [reply include/evmc/evmc.h:530]: What is this rule 2 of EIP-2200 referenced in the doc? "If _current value_ equals _new value_ (this is a no-op), `SLOAD_GAS` is deducted."? Or something else?

Comment (chfast) [reply include/evmc/evmc.h:530]: Oh, this is a mistake. This a group of this rule (2) and 3.2 without matching any subrules of 3.2. If you can parse my DSL, this can be e.g. `X → Y → Z`. Because these have the same cost, I decided to use single enum value for it.

If you think this is too confusing, I can assign separate values to these. But this suggests these have different behavior what is not the case.

Comment (yperbasis) [reply include/evmc/evmc.h:530]: I'd prefer separate `EVMC_STORAGE_UNCHANGED` and `EVMC_STORAGE_MODIFIED_AGAIN` for extra clarity, even if they are priced the same.

Comment (chfast) [reply include/evmc/evmc.h:530]: I reviewed my work on this and distinguishing these two requires additional check `current == new` (not needed to distinguish all other cases). Moreover, this would be inconvenient to test: e.g. state tests are not able to spot a mistake.

So I'm against.

Comment (yperbasis) [reply include/evmc/evmc.h:530]: OK, then please amend the comment to say that it covers both rule 2 as well as 3.2 without any of the sub-rules.

Comment (chfast) [reply include/evmc/evmc.h:532]: Please see the changed version.

Comment (yperbasis) [reply include/evmc/evmc.h:532]: What about calling this guy `EVMC_STORAGE_NO_OP`?

File: include/evmc/evmc.h [hunk 3]
```diff
@@ -502,36 +502,114 @@ typedef evmc_bytes32 (*evmc_get_storage_fn)(struct evmc_host_context* context,
...
+     * minimal gas cost.
+     * 0|X   -> 0 -> 0 (current value unchanged)
+     * 0|X|Y -> Y -> Y (current value unchanged)
+     * 0|X   -> Y -> Z (modified previously added/modified value)
+     *
+     * This is "catch all remaining" status. I.e. if all other statuses are correctly matched
+     * this status should be assigned to all remaining cases.
+     *
+     * This corresponds to the EIP-2200 rule 2 or rule 3.2 without sub-rules, where:
+     * c == v or (c != v, o != c, (o == 0 or (o != 0, c != 0, v != 0)), o != v)
```
Comment (gumb0) [inline include/evmc/evmc.h:530]: This is hard to parse, and rule numbers like `3.2.1.1` are not useful because don't exist in EIP.

I perfectly understand notation like `X -> Y -> 0`, why do you need this new one?

Comment (chfast) [reply include/evmc/evmc.h:530]: This is redundant. I try to explain what EIP rules it matches with the ordered list on condition to reach the particular rules in the spec tree in the EIP.

Any ideas how to improve this?

File: include/evmc/evmc.h (hunk 1 above)
Comment (chfast) [reply include/evmc/evmc.h:532]: I think `NO_OP` may be ok, however the spec uses this term in `current == value` context only.

Another possible name can be `UNAFFECTED`.

Comment (yperbasis) [reply include/evmc/evmc.h:532]: `UNAFFECTED` sounds good.

Comment (chfast) [reply include/evmc/evmc.h:532]: `REASSIGNED`?

Time for twitter pool.

File: include/evmc/evmc.h (hunk 3 above)
Comment (chfast) [reply include/evmc/evmc.h:530]: Do you think it is better to remove it? I can have separate doxygen chapter (md document) that describes all matching details using big tables.

File: include/evmc/evmc.h (hunk 1 above)
Comment (yperbasis) [reply include/evmc/evmc.h:532]: `REASSIGNED` is fine with me.

File: docs/storage_change_status.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
+# Storage change status
+
+> Analysis of possible storage change statuses in [EIP-2200].
+
+## Specification
+
+> This is the [EIP-2200] specification with modifications:
+> - the clause tree has been converted to ordered lists to be referencable,
```
Comment (gumb0) [inline docs/storage_change_status.md:8]:
```suggestion
> - the clause tree has been converted to ordered lists to be referenceable,
```

Comment (chfast) [reply docs/storage_change_status.md:9]: CLion says otherwise.

File: include/evmc/evmc.h (hunk 1 above)
Comment (chfast) [reply include/evmc/evmc.h:535]: Based on the comments from @gumb0 and my preferences, can we leave it as `EVMC_STATUS_MODIFIED_AGAIN`? Also documentation is updated.

**Comment by chfast** [comment]
Documentation rendered:
- https://output.circle-artifacts.com/output/job/df5edcb0-85f0-4745-87b8-a464456a4c34/artifacts/0/docs/storagestatus.html
- https://output.circle-artifacts.com/output/job/df5edcb0-85f0-4745-87b8-a464456a4c34/artifacts/0/docs/group__EVMC.html#gae012fd6b8e5c23806b507c2d3e9fb1aa

**Review by yperbasis — APPROVED** [review]

File: include/evmc/evmc.h (hunk 1 above)
Comment (chfast) [reply include/evmc/evmc.h]: Changed to `EVMC_STORAGE_ASSIGNED`.
